pkgs/development/tools/jdt-language-server/default.nix

```diff
@@ -0,0 +1,99 @@
+{ lib
+, stdenv
+, fetchurl
+, makeWrapper
+, jdk
+}:
+
+stdenv.mkDerivation rec {
+  pname = "jdt-language-server";
+  version = "1.7.0";
+  timestamp = "202112161541";
+
+  src = fetchurl {
+    url = "https://download.eclipse.org/jdtls/milestones/${version}/jdt-language-server-${version}-${timestamp}.tar.gz";
+    sha256 = "0ll5rgd8i8wfg2zz0ciapakl66qqaw344129qj72cyiixkgjh31g";
+  };
+
+  sourceRoot = ".";
+
+  buildInputs = [
+    jdk
+  ];
+
+  nativeBuildInputs = [
+    makeWrapper
+  ];
+
+  installPhase =
+    let
+      # The application ships with config directories for linux and mac
+      configDir = if stdenv.isDarwin then "config_mac" else "config_linux";
+      # The application will store it's data here. Note especially the escaping so
+      # that the env vars are interpreted at runtime and not during installPhase.
+      runtimePath = "\\\${XDG_CACHE_HOME:-\\$HOME/.cache}/jdt-language-server";
```

Comment:
Creating a runtime path is unnecessary.
The following flag should be enough to prevent jdtls from writing to the nix store:

```
-Dosgi.sharedConfiguration.area.readOnly=true
```

Reply:
@MrcJkb Thanks for the tip I was not aware. This solves for the option `-configuration "${runtimePath}/config"` and I could instead do `-configuration $out/share/config`, but it does not solve for `-data "${runtimePath}/data$PWD"`. How would you recommend storing the per-project runtime data?

Reply:
@matt-snider I don't think specifying where to store the per-project runtime data should be the responsibility of the launcher script.
I don't know how it is done in VSCode. But in NeoVim, I specify the `-data` directory based on the current working directory (see [the following example](https://github.com/mfussenegger/nvim-jdtls#data-directory-configuration)).

Reply:
@MrcJkb I agree with you on this and I think making the `-data` parameter a caller's concern is the right choice. However, the readonly configuration does not work. I added the option you suggested as well as several others mentioned [here](https://github.com/eclipse/eclipse.jdt.ls/issues/1823#issuecomment-1015422142). JDT LS still attempts to write logs to the config location:

```
The configuration area at '/nix/store/1dsx8w2aikc07209iymwlmwv9z5d4x1d-jdt-language-server-1.8.0/share/config' is not writable.  Please choose a writable location using the '-configuration' command line option.
```

There is also [this issue](https://github.com/eclipse/eclipse.jdt.ls/issues/1097).

If you still think this should work, perhaps you could try it out locally and send me the invocation that works.

Reply:
See [most recent change](https://github.com/NixOS/nixpkgs/compare/a87a0389497da96ac14fc927a0306bdfb4aca6a8..02859e85bad35b5ec1832fc8008d7ffdec0e279d) where I moved `-data` to a caller-specified option

Reply:
@matt-snider that's strange.
I have a package built from the milestone binaries [in my NUR repo](https://github.com/MrcJkb/nur-packages/blob/master/pkgs/eclipse-jdt-language-server/derivation.nix) that works fine.
It does have some additional flags that are not set in your launcher, namely:

```
-Dosgi.checkConfiguration=true 
-Dosgi.configuration.cascaded=true
```
And I use the `-Dosgi.sharedConfiguration.area` option instead of the `--configuration` flag.
Some of the flags you have set are missing in my wrapper, too (e.g. the module path configurations and the log level).
I don't have time today, but I hope to check out your branch this weekend and play around with it.

Reply:
@MrcJkb Ahh that works! Thank you for the tip that's much cleaner now. You can see the change [here](https://github.com/NixOS/nixpkgs/compare/02859e85bad35b5ec1832fc8008d7ffdec0e279d..0202ad1ef92e385a5445092239ad1f1e1dc2cbf3).

The problem was that I was setting all of the flags you did including `-Dosgi.sharedConfiguration.area=$out/share/config` but didn't leave out `-configuration`.

It would be great if you and any others could test out the derivation and make sure it still works properly for their use cases @Jumziey @winston0410

Reply:
@matt-snider I'm a week late, but I just installed it from your branch, and it works for me in NeoVim.